### ISSUE TITLE
[DataGrid] Column virtualization should not use index as key

### ISSUE THREAD

**Issue by andreasheim** [mui-org/material-ui-x#3498]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

https://github.com/mui-org/material-ui-x/pull/2673/files

`key` was changed to be an index: 
<img width="898" alt="Screen Shot 2021-12-16 at 11 07 35 AM" src="https://user-images.githubusercontent.com/399306/147167639-3bf45819-6fe6-45a1-b163-ab338bc168f4.png">

Where `i` is define as 
```
for (let i = 0; i < renderedColumns.length; i += 1) {
```

While the purpose per comment is to recycle cells, that's not how react tracks components. If you scroll content horizontally, you want react to render the contents of the cell into the same component. In order to do so, the key needs to stay the same as cells shift left or right. The key needs to be based on the content of the cell, not the index. 

This might be able to be observable by focusing a cell. E.g. if you focus the last visible cell to the right, then start scrolling to the right, it should remain focused as the content becomes 2nd to last. 

However when using index as key, focus will jump to the next cell coming into view. 

I haven't tried to repro this here, however this is something I've experienced often before when using index as key. 

### Expected behavior 🤔

Keys should be content based, not index based. 

### Steps to reproduce 🕹

Steps:

1.
2.
3.
4.

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by m4theushw** [comment]
This was already fixed by #3483 

https://github.com/mui-org/material-ui-x/blob/25a374d2408f34b2d715cd9b6ba923a74f5144aa/packages/grid/_modules_/grid/components/GridRow.tsx#L208

**Comment by andreasheim** [comment]
Perfect, thank you. then this can be closed.
